### PR TITLE
Feature/time based order

### PR DESCRIPTION
**Pull request by davidvuong**
This PR adjusts the existing next-price order component in `FuturesV2` such that the execution of orders are based on time rather than a semi-arbitrary price update from CL feeds. A trader can specify a `desiredTimeDelta` in `submitDelayedOrder`, allowing their order to be filled after `block.timestamp + desiredTimeDelta`.

At the same time, order can also be filled if in the situation where a CL price update occurs before `block.timestamp + desiredTimeDelta` (denoted as `executableAtTime` in the `DelayedOrder` struct).

For validation and additional configurability, each market has two additional parameters, `minDelayTimeDelta` and `maxDelayTimeDelta`. When submitting an order, the `desiredTimeDelta` must be between the lower and upper bound respectively. 

All references to `nextPrice` has been renamed to `delayedOrder`, including documentation (at least I hope). Note that `nextPriceConfirmWindow` has not changed. 

An additional parameter (named `delayedOrderConfirmationWindow`) has been added to work in conjunction `nextPriceConfirmationWindow`. Cancellations and executions incorporate a time based window which allows orders to be cancelled and executables preventable depending on whether they are over or within the confirmation window, respectively.

Note: Majority of contract changes can be found in `contracts/FuturesV2MarketDelayedOrders.sol`.

### CODE REVIEW

File: contracts/MixinFuturesV2DelayedOrders.sol [hunk 1]
```diff
@@ -10,38 +10,53 @@ import "./FuturesV2MarketMutations.sol";
...
  sacrifices to the UX / risk of the traders (e.g. blocking all actions, or penalizing failures too much).
  */
-contract MixinFuturesV2NextPriceOrders is FuturesV2MarketMutations {
-    /// @dev Holds a mapping of accounts to orders. Only one order per account is supported
-    mapping(address => NextPriceOrder) public nextPriceOrders;
-
-    ///// Mutative methods
+contract MixinFuturesV2DelayedOrders is FuturesV2MarketMutations {
+    /// Minimum amount of time (in seconds) before a delayed order can be executed.
+    uint public constant MIN_ORDER_DELAY = 60 seconds;
```
Comment (leomassazza) [inline contracts/MixinFuturesV2DelayedOrders.sol:21]: this constant should be a configuration parameter. See MixinFuturesV2MarketSettings.sol / FuturesV2MarketSettings.sol

File: contracts/MixinFuturesV2DelayedOrders.sol [hunk 2]
```diff
@@ -10,38 +10,53 @@ import "./FuturesV2MarketMutations.sol";
...
     }
 
-    /// same as submitNextPriceOrder but emits an event with the tracking code
+    /// same as submitDelayedOrder but emits an event with the tracking code
     /// to allow volume source fee sharing for integrations
-    function submitNextPriceOrderWithTracking(int sizeDelta, bytes32 trackingCode) external {
-        _submitNextPriceOrder(sizeDelta, trackingCode);
+    function submitDelayedOrderWithTracking(
+        int sizeDelta,
+        uint maxTimeDelta,
```
Comment (leomassazza) [inline contracts/MixinFuturesV2DelayedOrders.sol:44]: I'd remove the delay from the function spec and fix it to the min delay configured. (Already asked Afif for the product definition on this)

**Review by leomassazza — CHANGES_REQUESTED** [review]
love it! nice job. There's a fix needed in the cancellations (to prevent others to cancel the order before time) and maybe add a max delay config param to the order (see Afif comments on discord)

File: contracts/MixinFuturesV2DelayedOrders.sol (hunk 2 above)
Comment (davidvuong) [reply contracts/MixinFuturesV2DelayedOrders.sol:44]: We discussed this a little further on Discord and after a few rounds, Afif wanted to allow traders to specify a delay between a min and max (inclusive). As such, I have renamed everything to be more canonical; desired, min, and max.

File: contracts/MixinFuturesV2DelayedOrders.sol [hunk 3]
```diff
@@ -103,9 +127,9 @@ contract MixinFuturesV2NextPriceOrders is FuturesV2MarketMutations {
      *  or send to the msg.sender if it's not the account holder.
      * @param account the account for which the stored order should be cancelled
      */
-    function cancelNextPriceOrder(address account) external {
+    function cancelDelayedOrder(address account) external {
```
[the remark this answers is not in the document]
Comment (davidvuong) [reply contracts/MixinFuturesV2DelayedOrders.sol:130]: oops! Good catch!

**Review by leomassazza — APPROVED** [review]

**Comment by leomassazza** [comment]
Approved, but please, don't merge until futuresV2 is audited to ease the reviews

File: contracts/FuturesV2MarketDelayedOrders.sol [hunk 4]
```diff
@@ -31,32 +31,53 @@ contract FuturesV2MarketNextPriceOrders is IFuturesV2MarketNextPriceOrders, Futu
...
+        uint desiredTimeDelta,
+        bytes32 trackingCode
+    ) internal {
         // check that a previous order doesn't exist
-        require(marketState.nextPriceOrders(messageSender).sizeDelta == 0, "previous order exists");
+        require(marketState.delayedOrders(messageSender).sizeDelta == 0, "previous order exists");
+
+        // ensure the desiredTimeDelta is above the minimum required delay
+        bytes32 marketKey = marketState.marketKey();
+        require(
```
Comment (leomassazza) [inline contracts/FuturesV2MarketDelayedOrders.sol:77]: According to the SIP,` desiredTimeDelta == 0` is a valid value and defaults to `_minDelayTimeDelta`

**Comment by davidvuong** [comment]
Builds are failing due to an `npm audit` error. Awaiting fix from `futuresV2`. This is the same problem in #1906. While `futuresV2` is still in development, I'll be merging (not rebasing) commits and resolving tests to avoid drifting too far. Same applies to #1906.

```
npx hardhat test test/contracts/PerpsV2Market.js

  165 passing (4m)
  31 pending
```

```
npx hardhat test test/contracts/PerpsV2Market.delayedOrder.js

  41 passing (1m)
```

File: contracts/PerpsV2MarketDelayedOrders.sol [hunk 5]
```diff
@@ -259,14 +259,20 @@ contract PerpsV2MarketDelayedOrders is IPerpsV2MarketDelayedOrders, PerpsV2Marke
         // emit event for modifying the position (refunding fee/s)
         emitPositionModified(position.id, account, position.margin, position.size, 0, currentPrice, fundingIndex, 0);
 
-        // the correct price for the past round
-        (uint pastPrice, ) = _exchangeRates().rateAndTimestampAtRound(marketState.baseAsset(), order.targetRoundId);
+        // price depends on whether the delay or price update has reached/occurred first
+        uint executePrice = currentPrice;
+        if (currentRoundId > order.targetRoundId) {
```
Comment (leomassazza) [inline contracts/PerpsV2MarketDelayedOrders.sol:264]: shouldn't be (currentRoundId >= order.targetRoundId) ? (I didn't go to border cases but at first sight it looks >= is more accurate than strictly greater.

**Pull request merged**